server/src/main/java/org/opensearch/index/store/remote/file/OnDemandCompositeBlockIndexInput.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.store.remote.file;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.store.FSDirectory;
+import org.apache.lucene.store.IOContext;
+import org.apache.lucene.store.IndexInput;
+import org.opensearch.common.lucene.store.InputStreamIndexInput;
+import org.opensearch.index.store.RemoteSegmentStoreDirectory;
+import org.opensearch.index.store.remote.filecache.FileCache;
+import org.opensearch.index.store.remote.utils.BlobFetchRequest;
+import org.opensearch.index.store.remote.utils.BlockIOContext;
+import org.opensearch.index.store.remote.utils.TransferManager;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * OnDemandCompositeBlockIndexInput is used by the Composite Directory to read data in blocks from Remote and cache those blocks in FileCache
+ */
+public class OnDemandCompositeBlockIndexInput extends OnDemandBlockIndexInput {
+
+    private static final Logger logger = LogManager.getLogger(OnDemandCompositeBlockIndexInput.class);
+    private final RemoteSegmentStoreDirectory remoteDirectory;
+    private final String fileName;
+    private final Long originalFileSize;
+    private final FSDirectory localDirectory;
+    private final IOContext context;
+    private final FileCache fileCache;
+    private final TransferManager transferManager;
+
+    public OnDemandCompositeBlockIndexInput(
+        RemoteSegmentStoreDirectory remoteDirectory,
```

Comment:
I think this component should take in the metadata required and a transfer manager instead of the directory.  With `OnDemandBlockSnapshotIndexInput` that metadata is FileInfo.  Would prefer we pass the metadata this class needs instead with remote store or reuse OnDemandBlockSnapshotIndexInput directly.